src/components/TilePicker.tsx

```diff
@@ -0,0 +1,57 @@
+import React from 'react';
+import * as db from "../firebase";
+import { useForm } from "react-hook-form";
+
+export function TilePicker(props: {
+  tileName: string,
+  tileInfo: db.TileInfo,
+  map: db.MapRepr,
+  tiles: db.TileDict,
+  selectedRow?: number,
+  selectedCol?: number,
+  gameId: string
+}) {
+  const { register, handleSubmit } = useForm<{tileName: string}>();
+
+  const tileList = [...Object.keys(props.tiles)]
+  const tiles = tileList.map( (tileName) =>
+    <option key={tileName}>{tileName}</option>
+  )
+
+  const onSubmit = handleSubmit(async ({tileName}) => {
+    changeTile(tileName);
+  });
+
+  function changeTile(tileName: string) {
+    if (!tileList.includes(tileName)) {
+      createTile(tileName);
+    } else {
+      const row = props.selectedRow;
+      const col = props.selectedCol;
+      if (row !== undefined && col !== undefined) {
+        props.map[row][col] = tileName;
+        const map = props.map.flat();
+        db.UpdateGridForGame(props.gameId, map);
+      }}
+  }
+
+  function createTile(tileName: string) {
+    db.AddTile(tileName, props.tileInfo);
+  }
+
+  return (
+    <form className="tile-select-form" onSubmit={onSubmit}>
+      <label>Tile:
+        <input
```

Comment:
Using a single input that can be typed into is confusing - it signifies to me that you are changing the name of the tile. I think a <select> with a "Create New Tile" option would be more intuitive.